measurement/plugins/webpage_download/measurements.py

```diff
@@ -0,0 +1,172 @@
+        except ConnectionError as e:
+            return self._get_webpage_error("web-get", traceback=str(e))
+        except requests.exceptions.ReadTimeout as e:
+            return self._get_webpage_error("web-timeout", traceback=str(e))
+        try:
+            to_download = self._parse_html(r.text)
+        except TypeError as e:
+            return self._get_webpage_error("web-parse-rel", traceback=str(e))
+        try:
+            asset_download_metrics = self._download_assets(s, to_download, host)
+        except TypeError as e:
+            return self._get_webpage_error("web-assets", traceback=str(e))
+
+        primary_download_size = len(r.text)
+        asset_download_size = asset_download_metrics["asset_download_size"]
+        elapsed_time = asset_download_metrics["completion_time"] - start_time
+        download_rate = (primary_download_size + asset_download_size) * 8 / elapsed_time
+        failed_asset_downloads = asset_download_metrics["failed_asset_downloads"]
+
+        return WebpageMeasurementResult(
+            id=self.id,
+            url=url,
+            download_rate=download_rate,
+            download_rate_unit=NetworkUnit("bit/s"),
+            download_size=primary_download_size + asset_download_size,
+            download_size_unit=StorageUnit("B"),
+            asset_count=len(to_download),
+            failed_asset_downloads=failed_asset_downloads,
+            elapsed_time=elapsed_time,
+            elapsed_time_unit=TimeUnit("s"),
+            errors=[],
+        )
+
+    def _parse_html(self, content):
+        soup = BeautifulSoup(content, "html.parser")
+        # hlinks = soup.find_all("a")
+        imgs = soup.find_all("img")
+        links = soup.find_all("link")
+        scripts = soup.find_all("script")
+        to_download = []
+        for img in imgs:
+            if img.has_attr("src"):
+                to_download.append(img["src"])
+        for script in scripts:
+            if script.has_attr("src"):
+                to_download.append(script["src"])
+        for link in links:
+            if link.has_attr("rel") & link.has_attr("href"):
+                # Join in the case where `rel` is more than one word
+                rel = " ".join(link["rel"])
+                if rel in VALID_LINK_REL_ATTRIBUTES:
+                    to_download.append(link["href"])
+
+        return to_download
+
+    def _download_assets(self, session, to_download, host):
+        # Store the amount of bytes downloaded
+        asset_download_sizes = []
+        failed_asset_downloads = 0
+        for asset in to_download:
+            try:
+                # Identify data URLs (already downloaded inline, counted in main download size)
+                if "data:" in asset:
+                    continue
+
+                # Check if path w/o preceeding slashes is a valid URL
+                if asset.startswith("//"):
+                    download_url = "https:" + asset
```

Comment:
Are we assuming that every asset will be ssl-secured here? What happens if an asset is not ssl-secured? Shouldn't we be using `http`|`https` based on the `asset` url or at least the `url` originally passed to the measurement?